### PR TITLE
Remove unnecessary function literal creations

### PR DESCRIPTION
**Pull request by zku**
Remove some unnecessary function literal creations (major contributor to small GC allocations -> sweeps). A lot of time is spent allocating and cleaning up after all the new function literals being created, and some of these don't really have to exist, i.e. a pattern like

```OnAction = func(sim *Simulation) { Foo(sim) }```

could just be expressed as

```OnAction = Foo```

without the additional allocation (this is already used in most places luckily). Over 100k iterations, it seems like we spend most of our time creating and re-scheduling `PendingAction`s, which all create these unnecessary function literals repeatedly.

Using the raid bench with 1000 iterations and 300s duration over 10 repetitions yields a ~25% allocs/op reduction with minor performance gains due to GC activity reductions (~5% faster). Of course this will be very machine specific, but reducing small unnecessary allocations should be an improvement under most circumstances.

Lastly, also changes the unit test to bind to localhost (vs. all interfaces) to prevent Windows UAC prompts when running the unit tests.

For reference:

```
goos: windows
goarch: amd64
pkg: github.com/wowsims/wotlk/sim
cpu: AMD Ryzen 7 5800X 8-Core Processor
```

### CODE REVIEW

File: sim/core/periodic_action.go [hunk 1]
```diff
@@ -15,21 +15,12 @@ type DelayedActionOptions struct {
 }
 
 func NewDelayedAction(sim *Simulation, options DelayedActionOptions) *PendingAction {
-	pa := &PendingAction{
```
Comment (vigo2) [inline sim/core/periodic_action.go:18]: it's simpler to panic() if `options.OnAction` is nil here...

File: sim/core/periodic_action.go [hunk 2]
```diff
@@ -61,12 +52,15 @@ func NewPeriodicAction(sim *Simulation, options PeriodicActionOptions) *PendingA
 	pa := &PendingAction{
```
Comment (vigo2) [inline sim/core/periodic_action.go:52]: ... and here as well. This allows to keep the implicit "PendingAction.OnAction() must not be nil" invariant elsewhere, and makes it more explicit.
The code changes become very minimalistic then, too :)

File: sim/core/gcd.go [hunk 3]
```diff
@@ -15,9 +15,7 @@ func (unit *Unit) newHardcastAction(sim *Simulation) {
 	if unit.hardcastAction == nil {
 		pa := &PendingAction{
 			NextActionAt: unit.Hardcast.Expires,
-			OnAction: func(sim *Simulation) {
-				// Don't need to do anything, the Advance() call will take care of the hardcast.
-			},
+			// No action, the Advance() call will take care of the hardcast.
```
Comment (vigo2) [inline sim/core/gcd.go:18]: You'd have to leave this "as is" then, which isn't much of a problem, as this case is very rarely taken. The main culprits are NewDelayedAction() and NewPeriodicAction().
Alternatively, core could add a `func DoNothing(*Simulation) {}`, but that's probably overkill.

**Comment by vigo2** [comment]
Otherwise, LGTM. Thanks for doing this!

**Review by vigo2 — CHANGES_REQUESTED** [review]

File: sim/core/periodic_action.go (hunk 2 above)
Comment (zku) [reply sim/core/periodic_action.go:60]: Done!

File: sim/core/periodic_action.go (hunk 1 above)
Comment (zku) [reply sim/core/periodic_action.go:18]: Done!

File: sim/core/gcd.go (hunk 3 above)
Comment (zku) [reply sim/core/gcd.go]: Reverted.

**Review by lologarithm — APPROVED** [review]

**Comment by zku** [comment]
Can you please have another look @vigo2 ?

**Review by vigo2 — APPROVED** [review]

**Pull request merged**